File: src/main/java/com/fifo/ticketing/domain/performance/mapper/PerformanceMapper.java
package com.fifo.ticketing.domain.performance.mapper;

import com.fifo.ticketing.domain.performance.dto.AdminPerformanceDetailResponse;
import com.fifo.ticketing.domain.performance.dto.AdminPerformanceResponseDto;
import com.fifo.ticketing.domain.performance.dto.LikedPerformanceDto;
import com.fifo.ticketing.domain.performance.dto.PerformanceDetailResponse;
import com.fifo.ticketing.domain.performance.dto.PerformanceRequestDto;
import com.fifo.ticketing.domain.performance.dto.PerformanceResponseDto;
import com.fifo.ticketing.domain.performance.dto.PerformanceSeatGradeDto;
import com.fifo.ticketing.domain.performance.entity.Grade;
import com.fifo.ticketing.domain.performance.entity.Performance;
import com.fifo.ticketing.domain.performance.entity.Place;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import org.springframework.data.domain.Page;

public class PerformanceMapper {

    private PerformanceMapper() {
    }

    public static PerformanceDetailResponse toDetailResponseDto(Performance performance,
        List<PerformanceSeatGradeDto> seatGrades, String urlPrefix) {
        return PerformanceDetailResponse.builder()
            .performanceId(performance.getId())
            .title(performance.getTitle())
            .description(performance.getDescription())
            .category(performance.getCategory().name())
            .startTime(performance.getStartTime())
            .encodedFileName(performance.getFile().getEncodedFileName())
            .endTime(performance.getEndTime())
            .reservationStartTime(performance.getReservationStartTime())
            .placeName(performance.getPlace().getName())
            .address(performance.getPlace().getAddress())
            .performanceStatus(performance.isPerformanceStatus())
            .totalSeats(performance.getPlace().getTotalSeats())
            .seatGrades(seatGrades)
            .urlPrefix(urlPrefix)
            .build();
    }

    public static AdminPerformanceDetailResponse toAdminDetailResponseDto(Performance performance,
        List<PerformanceSeatGradeDto> seatGrades, String urlPrefix) {
        return new AdminPerformanceDetailResponse(
                performance.getId(),
                performance.getTitle(),
                performance.getDescription(),
                performance.getPlace().getName(),
                performance.getFile().getEncodedFileName(),
                performance.getPlace().getAddress(),
                performance.getCategory().name(),
                performance.getStartTime(),
                performance.getEndTime(),
                performance.isDeletedFlag(),
                performance.isPerformanceStatus(),
                performance.getPlace().getTotalSeats(),
                seatGrades,
                urlPrefix
        );
    }

    public static PerformanceSeatGradeDto toSeatGradeDto(Grade grade) {
        return new PerformanceSeatGradeDto(grade.getGrade(), grade.getDefaultPrice(),
                grade.getSeatCount());
    }

    public static PerformanceResponseDto toPerformanceResponseDto(Performance performance,
        String urlPrefix) {
        return PerformanceResponseDto.builder()
            .id(performance.getId())
            .encodedFileName(performance.getFile().getEncodedFileName())
            .title(performance.getTitle())
            .category(performance.getCategory().name())
            .place(performance.getPlace().getName())
            .startTime(performance.getStartTime())
            .endTime(performance.getEndTime())
            .reservationStartTime(performance.getReservationStartTime())
            .performanceStatus(performance.isPerformanceStatus())
            .urlPrefix(urlPrefix)
            .build();
    }

    public static Performance toEntity(PerformanceRequestDto dto, Place place) {
        LocalDate dateOnly = dto.getReservationStartTime().toLocalDate();
        LocalDateTime fixedReservationStartTime = dateOnly.atTime(13, 0);
        return Performance.builder()
            .title(dto.getTitle())
            .description(dto.getDescription())
            .place(place)
            .startTime(dto.getStartTime())
            .endTime(dto.getEndTime())
            .category(dto.getCategory())
            .performanceStatus(dto.isPerformanceStatus())
            .reservationStartTime(fixedReservationStartTime)
            .build();
    }

    public static Page<PerformanceResponseDto> toPagePerformanceResponseDto(
        Page<Performance> performances, String urlPrefix) {
        return performances.map(
            performance -> PerformanceMapper.toPerformanceResponseDto(performance, urlPrefix));
    }

    public static AdminPerformanceResponseDto toAdminPerformanceResponseDto(Performance performance,
        String urlPrefix) {
        return new AdminPerformanceResponseDto(
            performance.getId(),
            performance.getFile().getEncodedFileName(),
            performance.getTitle(),
            performance.getDescription(),
            performance.getCategory().name(),
            performance.getPlace().getName(),
            performance.getStartTime(),
            performance.getEndTime(),
            performance.getReservationStartTime(),
            performance.isPerformanceStatus(),
            urlPrefix
        );
    }

    public static LikedPerformanceDto toLikedPerformanceDto(Performance performance,
        String prefix) {
        return new LikedPerformanceDto(
            performance.getId(),
            performance.getTitle(),
            performance.getFile().getEncodedFileName(),
            performance.getStartTime(),
            performance.getEndTime(),
            performance.getPlace().getName(),
            prefix
        );
    }

    public static Page<AdminPerformanceResponseDto> toPageAdminPerformanceResponseDto(
        Page<Performance> performances, String urlPrefix) {
        return performances.map(
            performance -> PerformanceMapper.toAdminPerformanceResponseDto(performance, urlPrefix));
    }

    public static Page<LikedPerformanceDto> toPageLikedPerformanceDto(
        Page<Performance> performances, String urlPrefix) {
        return performances.map(
            performance -> PerformanceMapper.toLikedPerformanceDto(performance, urlPrefix)
        );
    }
}

File: src/main/java/com/fifo/ticketing/domain/performance/mapper/PerformanceMapper.kt
package com.fifo.ticketing.domain.performance.mapper

import com.fifo.ticketing.domain.performance.dto.*
import com.fifo.ticketing.domain.performance.entity.Grade
import com.fifo.ticketing.domain.performance.entity.Performance
import com.fifo.ticketing.domain.performance.entity.Place
import org.springframework.data.domain.Page

object PerformanceMapper {
    @JvmStatic
    fun toDetailResponseDto(
        performance: Performance,
        seatGrades: List<PerformanceSeatGradeDto>, urlPrefix: String
    ): PerformanceDetailResponse {
        return PerformanceDetailResponse(
            performance.id!!,
            performance.title,
            performance.description,
            performance.place.name,
            performance.file!!.encodedFileName,
            performance.place.address,
            performance.category.name,
            performance.startTime,
            performance.endTime,
            performance.reservationStartTime,
            performance.performanceStatus,
            performance.place.totalSeats,
            seatGrades,
            urlPrefix
        )
    }

    @JvmStatic
    fun toSeatGradeDto(grade: Grade): PerformanceSeatGradeDto {
        return PerformanceSeatGradeDto(
            grade.grade, grade.defaultPrice,
            grade.seatCount
        )
    }

    @JvmStatic
    fun toPerformanceResponseDto(
        performance: Performance,
        urlPrefix: String
    ): PerformanceResponseDto {
        return PerformanceResponseDto(
            performance.id!!,
            performance.file!!.encodedFileName,
            performance.title,
            performance.category.name,
            performance.place.name,
            performance.startTime,
            performance.endTime,
            performance.reservationStartTime,
            performance.performanceStatus,
            urlPrefix
        )
    }

    @JvmStatic
    fun toEntity(dto: PerformanceRequestDto, place: Place): Performance {
        val dateOnly = dto.reservationStartTime.toLocalDate()
        val fixedReservationStartTime = dateOnly.atTime(13, 0)
        return Performance(
            null,
            dto.title,
            dto.description,
            place,
            dto.startTime,
            dto.endTime,
            dto.category,
            dto.performanceStatus,
            false,
            fixedReservationStartTime,
            null
        )
    }

    @JvmStatic
    fun toPagePerformanceResponseDto(
        performances: Page<Performance>, urlPrefix: String
    ): Page<PerformanceResponseDto> {
        return performances.map { performance: Performance ->
            toPerformanceResponseDto(
                performance,
                urlPrefix
            )
        }
    }

    @JvmStatic
    fun toLikedPerformanceDto(
        performance: Performance,
        prefix: String
    ): LikedPerformanceDto {
        return LikedPerformanceDto(
            performance.id!!,
            performance.title,
            performance.file!!.encodedFileName,
            performance.startTime,
            performance.endTime,
            performance.place.name,
            prefix
        )
    }

    @JvmStatic
    fun toPageLikedPerformanceDto(
        performances: Page<Performance>, urlPrefix: String
    ): Page<LikedPerformanceDto> {
        return performances.map { performance: Performance ->
            toLikedPerformanceDto(
                performance,
                urlPrefix
            )
        }
    }
}
